src/SDL_properties.c

```diff
@@ -0,0 +1,243 @@
+        SDL_UnlockRWLock(SDL_properties_lock);
+    }
+    if (!properties) {
+        SDL_InvalidParamError("props");
+        return -1;
+    }
+    return SDL_LockMutex(properties->lock);
+}
+
+void SDL_UnlockProperties(SDL_PropertiesID props)
+{
+    SDL_Properties *properties = NULL;
+
+    if (SDL_LockRWLockForReading(SDL_properties_lock) == 0) {
+        SDL_FindInHashTable(SDL_properties, (const void *)(uintptr_t)props, (const void **)&properties);
+        SDL_UnlockRWLock(SDL_properties_lock);
+    }
+    if (!properties) {
+        return;
+    }
+    SDL_UnlockMutex(properties->lock);
+}
+
+int SDL_SetProperty(SDL_PropertiesID props, const char *name, void *value, void (*cleanup)(void *userdata, void *value), void *userdata)
+{
+    SDL_Properties *properties = NULL;
+    SDL_Property *property = NULL;
+
+    if (!name || !*name) {
+        return SDL_InvalidParamError("name");
+    }
+
+    if (SDL_LockRWLockForReading(SDL_properties_lock) == 0) {
+        SDL_FindInHashTable(SDL_properties, (const void *)(uintptr_t)props, (const void **)&properties);
+        SDL_UnlockRWLock(SDL_properties_lock);
+    }
+    if (!properties) {
+        return SDL_InvalidParamError("props");
+    }
+
+    SDL_RemoveFromHashTable(properties->props, name);
+    if (value) {
+        property = (SDL_Property *)SDL_malloc(sizeof(*property));
+        if (!property) {
+            SDL_OutOfMemory();
+            goto error;
+        }
+        property->value = value;
+        property->cleanup = cleanup;
+        property->userdata = userdata;
+        if (!SDL_InsertIntoHashTable(properties->props, SDL_strdup(name), property)) {
+            goto error;
+        }
+    }
+    /* All done! */
+    return 0;
+
+error:
+    SDL_FreeProperty(NULL, property, NULL);
+    return -1;
+}
+
+void *SDL_GetProperty(SDL_PropertiesID props, const char *name)
+{
+    SDL_Properties *properties = NULL;
+    SDL_Property *property = NULL;
+
+    if (!name || !*name) {
+        SDL_InvalidParamError("name");
+        return NULL;
+    }
+
+    if (SDL_LockRWLockForReading(SDL_properties_lock) == 0) {
+        SDL_FindInHashTable(SDL_properties, (const void *)(uintptr_t)props, (const void **)&properties);
+        SDL_UnlockRWLock(SDL_properties_lock);
+    }
+    if (!properties) {
+        SDL_InvalidParamError("props");
+        return NULL;
+    }
+    if (!SDL_FindInHashTable(properties->props, name, (const void **)&property)) {
+        SDL_SetError("Couldn't find property named %s", name);
+        return NULL;
+    }
+    return property->value;
+}
+
+void SDL_DestroyProperties(SDL_PropertiesID props)
+{
+    if (SDL_LockRWLockForWriting(SDL_properties_lock) == 0) {
+        SDL_RemoveFromHashTable(SDL_properties, (const void *)(uintptr_t)props);
+        SDL_UnlockRWLock(SDL_properties_lock);
+
+        /* If this was the last property, go ahead and clean up */
+        if (SDL_HashTableEmpty(SDL_properties)) {
```

Comment:
Do we want to shut everything down when we run out of properties? I'm thinking we should hook into SDL_Init/SDL_Quit instead, even if this means firing all the app's cleanup callbacks in a giant batch, so they aren't surprised by one after they've deinitialized SDL all their subsystems, and so we aren't destroying and creating the RWlock multiple times if there's a few properties floating in and out.

Reply:
Yeah, I debated that. I could see really bad things happening if these started firing during the shutdown process, but that's probably the right call for now.